src/pgpsda_fmt_plug.c

```diff
@@ -0,0 +1,236 @@
+/*
+ * Format for brute-forcing PGP SDAs (self-decrypting archives).
+ *
+ * This software is Copyright (c) 2017, Dhiru Kholia <dhiru.kholia at gmail.com>,
+ * and it is hereby released to the general public under the following terms:
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted.
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_pgpsda;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_pgpsda);
+#else
+
+#include <string.h>
+#include <openssl/cast.h>
+
+#include "arch.h"
+#include "misc.h"
+#include "memory.h"
+#include "common.h"
+#include "formats.h"
+#include "johnswap.h"
+#ifdef _OPENMP
+#include <omp.h>
+#ifndef OMP_SCALE
+#define OMP_SCALE               1 // this is a slow format
+#endif
+#endif
+#include "sha.h"
+#include "loader.h"
+#include "pgpsda_common.h"
+#include "memdbg.h"
+
+#define FORMAT_LABEL            "pgpsda"
+#define FORMAT_NAME             ""
+#define ALGORITHM_NAME          "PGP SDA SHA1 " ARCH_BITS_STR
+#define PLAINTEXT_LENGTH        125
+#define SALT_SIZE               sizeof(struct custom_salt)
+#define SALT_ALIGN              sizeof(int)
+#define BINARY_SIZE             8
+#define BINARY_ALIGN            sizeof(uint32_t)
+#define BENCHMARK_COMMENT       ""
+#define BENCHMARK_LENGTH        -1
+#define MIN_KEYS_PER_CRYPT      1
+#define MAX_KEYS_PER_CRYPT      1
+#define FORMAT_TAG              "$pgpsda$"
+#define FORMAT_TAG_LENGTH       (sizeof(FORMAT_TAG) - 1)
+
+static struct custom_salt *cur_salt;
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static uint32_t (*crypt_out)[BINARY_SIZE * 2 / sizeof(uint32_t)];
+
+static void init(struct fmt_main *self)
+{
+#ifdef _OPENMP
+	static int omp_t = 1;
+	omp_t = omp_get_max_threads();
+	self->params.min_keys_per_crypt *= omp_t;
+	omp_t *= OMP_SCALE;
+	self->params.max_keys_per_crypt *= omp_t;
+#endif
+	saved_key = mem_calloc(sizeof(*saved_key), self->params.max_keys_per_crypt);
+	crypt_out = mem_calloc(sizeof(*crypt_out), self->params.max_keys_per_crypt);
+}
+
+static void done(void)
+{
+	MEM_FREE(saved_key);
+	MEM_FREE(crypt_out)
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static void *get_binary(char *ciphertext)
+{
+	static union {
+		unsigned char c[BINARY_SIZE];
+		uint32_t dummy;
+	} buf;
+	unsigned char *out = buf.c;
+	char *p;
+	int i;
+	p = strrchr(ciphertext, '*') + 1;
+	for (i = 0; i < BINARY_SIZE; i++) {
+		out[i] = (atoi16[ARCH_INDEX(*p)] << 4) | atoi16[ARCH_INDEX(p[1])];
+		p += 2;
+	}
+
+	return out;
+}
+
+#undef SHA1_DIGEST_LENGTH
+#define SHA1_DIGEST_LENGTH 20
+
+static void pgpsda_kdf(char *password, unsigned char *salt, unsigned char *key)
+{
+	int plen;
+	int iterations = cur_salt->iterations;
+	SHA_CTX ctx; // SHA1 usage is hardcoded
+	uint16_t i = 0, j = 0;
```

Comment:
What is the point in using uint16_t instead of native integer? If anything, it's slower. Not that I expect it to be measurable.

Reply:
Good points!. The original code used a 16-bit unsigned number, so I used uint16_t. Since the iteration count seems to be fixed at 16,000 (no integer wrap-around stuff seems to be involved), using a regular int would have also worked.

The `iterations` variable is also of type uint16_t in the original code (and in the on-disk headers).

Reply:
In the latest commit, I have changed the type of these variables to a regular int.